### DISCUSSION TITLE
Repo-defined allowedPostUpgradeCommands ignored

### DISCUSSION THREAD

**Discussion by cjfinnell** [renovatebot/renovate#22716] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

In our attempts to use `postUpgradeTasks`, it appears as though the repository-defined `allowedPostUpgradeCommands` is being overwritten by the base github.com app defined fields; the logs show that:
1. Our repo-specified `allowedPostUpgradeCommands` and `postUpgradeTasks` sections are being parsed in the config-setup stage without issue (locally validated with `renovate-config-validator`)
2. The `postUpgradeTasks` are being attempted, but
3. The repo-defined `allowedPostUpgradeCommands` are *not* being respected at time of execution, erroring with `Post-upgrade task did not match any on allowedPostUpgradeCommands list`

These are the most relevant discussions I could find; both are `self-hosted` and appear to be unrelated to this issue:
1. https://github.com/renovatebot/renovate/discussions/21247
2. https://github.com/renovatebot/renovate/discussions/22102

I have not created a minimal reproduction, as I suspect the details (included) of our shared-base setup may be relevant to the failure.

---

Redacted shared base config (referenced as `github>redacted-private-org/renovate-config` in the redacted/attached logs):
```
{
  "extends": [
    "config:base",
    "schedule:earlyMondays"
  ],
  "enabledManagers": [
    "gomod",
    "github-actions",
    "circleci",
    "git-submodules"
  ],
  "hostRules": [
    {
      "hostType": "npm",
      "matchHost": "redacted",
      "encrypted": {
        "token": "redated"
      },
      "authType": "Bearer"
    }
  ],
  "npmrc": "@redacted",
  "postUpdateOptions": [
    "gomodTidy"
  ],
  "packageRules": [
    {
      "matchManagers": [
        "gomod"
      ],
      "matchUpdateTypes": [
        "minor",
        "patch",
        "digest"
      ],
      "matchDepTypes": [
        "require"
      ],
      "groupName": "all non-major Go dependencies",
      "groupSlug": "all-minor-patch-go"
    }
  ]
}
```

Logs pulled from `Job Id: e1172271-a1ad-48db-91e5-f12b3a28850b`.

### Relevant debug logs

<details><summary>Logs</summary>

Logs pulled from `Job Id: e1172271-a1ad-48db-91e5-f12b3a28850b`

```
DEBUG: Combined config
{
  "config": {
    "allowPostUpgradeCommandTemplating": true,
    "allowedPostUpgradeCommands": [
      "^git add --all$",
      "^git reset$",
      "^npx beachball change( --no-fetch)? --no-commit --type (patch|none) --message '{{{commitMessage}}}'$",
      "^pwd$"
    ],
    "baseDir": "/tmp/worker/e44442/ff009f",
    "dependencyDashboardFooter": "\n- [ ] <!-- manual job -->Check this box to trigger a request for Renovate to run again on this repository\n",
    "extends": [
      "github>whitesource/merge-confidence:beta"
    ],
    "forkProcessing": "disabled",
    "gitAuthor": "renovate[bot] <redacted+renovate[bot]@users.noreply.github.com>",
    "gitIgnoredAuthors": [
      "redacted+renovate[bot]@users.noreply.github.com"
    ],
    "hostRules": [
      {
        "hostType": "docker",
        "matchHost": "docker.io",
        "password": "***********",
        "username": "mdpprod2renovate"
      }
    ],
    "logContext": "redacted",
    "logFile": "redacted",
    "logFileLevel": "debug",
    "onboardingConfig": {
      "$schema": "https://docs.renovatebot.com/renovate-schema.json",
      "extends": [
        "config:base"
      ]
    },
    "onboardingNoDeps": false,
    "platform": "github",
    "platformCommit": true,
    "prHeader": "[![Mend Renovate](https://app.renovatebot.com/images/banner.svg)](https://renovatebot.com)",
    "repositories": [
      "redacted/repo"
    ],
    "token": "***********",
    "username": "renovate[bot]",
    "repositoryCache": "enabled",
    "repositoryCacheType": "s3://mend-developer-platform-prod/renovate/",
    "binarySource": "docker",
    "redisUrl": "redacted",
    "dockerChildPrefix": "renovate_a_",
    "dockerImagePrefix": "ghcr.io/containerbase",
    "privateKey": "***********",
    "privateKeyOld": "***********",
    "prFooter": "This PR has been generated by [Mend Renovate](https://www.mend.io/free-developer-tools/renovate/). View repository job log [here](https://app.renovatebot.com/dashboard#{{platform}}/{{repository}})."
  }
}

DEBUG: Found valid git version: 2.40.1
DEBUG: Using default github endpoint: https://api.github.com/
.
.
.
DEBUG: Repository config
{
  "fileName": ".github/renovate.json"
  "config": {
    "extends": [
      "github>redacted-private-org/renovate-config"
    ],
    "allowPostUpgradeCommandTemplating": true,
    "allowedPostUpgradeCommands": [
      "^make mocks$",
      "^make lint-fix$"
    ],
    "postUpgradeTasks": {
      "commands": [
        "make mocks",
        "make lint-fix"
      ],
      "executionMode": "branch"
    }
  }
}
DEBUG: Repo is onboarded
DEBUG: migrateAndValidate()
DEBUG: No config migration necessary
DEBUG: massaged config
{
  "config": {
    "extends": [
      "github>whitesource/merge-confidence:beta",
      "github>redacted-private-org/renovate-config"
    ],
    "allowPostUpgradeCommandTemplating": true,
    "allowedPostUpgradeCommands": [
      "^make mocks$",
      "^make lint-fix$"
    ],
    "postUpgradeTasks": {
      "commands": [
        "make mocks",
        "make lint-fix"
      ],
      "executionMode": "branch"
    }
  }
}

DEBUG: migrated config
{
  "config": {
    "extends": [
      "github>whitesource/merge-confidence:beta",
      "github>redacted-private-org/renovate-config"
    ],
    "allowPostUpgradeCommandTemplating": true,
    "allowedPostUpgradeCommands": [
      "^make mocks$",
      "^make lint-fix$"
    ],
    "postUpgradeTasks": {
      "commands": [
        "make mocks",
        "make lint-fix"
      ],
      "executionMode": "branch"
    }
  }
}
.
.
.

WARN: Post-upgrade task did not match any on allowedPostUpgradeCommands list (branch="renovate/all-minor-patch-go")
{
  "dep": "github.com/redacted-private-org/repo github.com/vektra/mockery/v2"
  "cmd": "make mocks"
  "allowedPostUpgradeCommands": [
    "^git add --all$",
    "^git reset$",
    "^npx beachball change( --no-fetch)? --no-commit --type (patch|none) --message '{{{commitMessage}}}'$",
    "^pwd$"
  ]
}

WARN: Post-upgrade task did not match any on allowedPostUpgradeCommands list (branch="renovate/all-minor-patch-go")
{
  "dep": "github.com/redacted-private-org/repo github.com/vektra/mockery/v2"
  "cmd": "make lint-fix"
  "allowedPostUpgradeCommands": [
    "^git add --all$",
    "^git reset$",
    "^npx beachball change( --no-fetch)? --no-commit --type (patch|none) --message '{{{commitMessage}}}'$",
    "^pwd$"
  ]
}
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment, marked as answer]
Post-upgrade commands are disabled by default on the hosted app due to security reasons. The list of allowed commands is always controlled in global config and never in repository config, so it is behaving by design

**Reply by cjfinnell** [reply]
Is this behavior documented anywhere as `self-hosted` only?

**Reply by cjfinnell** [reply]
I'm seeing that the option is defined in the [all-repository docs](https://docs.renovatebot.com/configuration-options/#postupgradetasks) *and* the [self-hosted only docs](https://docs.renovatebot.com/self-hosted-configuration/#allowedpostupgradecommands), definitely confusing.

**Reply by cjfinnell** [reply]
Nevermind, found the note:
> Post-upgrade tasks can only be used on self-hosted Renovate instances.

Thanks!

**Reply by rarkins** [reply]
> I'm seeing that the option is defined in the [all-repository docs](https://docs.renovatebot.com/configuration-options/#postupgradetasks) _and_ the [self-hosted only docs](https://docs.renovatebot.com/self-hosted-configuration/#allowedpostupgradecommands), definitely confusing.

You linked to two different option names - one is applicable for repo, and one for self-hosted config. So it's incorrect to say "the option" when it's two different options.

**Reply by cjfinnell** [reply]
Good catch, which begs the question: why is `postUpgradeTasks` included in the repo-applicable docs with `Note: Post-upgrade tasks can only be used on self-hosted Renovate instances` instead of simply being in the self-hosted docs?

**Reply by cjfinnell** [reply]
i.e. is this something I should clean up for a merge?
https://github.com/renovatebot/renovate/pull/22718

**Reply by rarkins** [reply]
The command is documented in the repo section because self-hosted users can then configure `postUpgradeTasks` in repo config. Repo config does not mean "hosted/app config", it means config which repos can configure themselves. You seem to be associating repo config = app config which is a misunderstanding. If we restricted `postUpgradeTasks` to global config only then individual repos couldn't configure their own.

**Reply by cjfinnell** [reply]
Gotcha, I was definitely operating under the assumption that there was one type of config, with `app config` being simply default values (having only played with the github.com hosted version). Thanks for the clarification!
